File: Scalar/live-sessions/Assignment/10 - prefix-sum/pick-from-both-sides.py
class Solution:
    def solve(self, A, B):
        """ 
        B = 3 ( should take three balls ) 
        L   |   R
        0   |   B           => 0 from Left and 3 from right # case 1
        1   |   B - 1       => 1 from Left and 2 from right # case 2
        2   |   B - 2       => 2 from left and 1 from right # case 3
        3   |   B - 3       => 3 from left and 0 from right # case 4

        """

        N = len(A)
        prefix = []
        prefix.append(A[0])
        for i in range(1, N):
            prefix.append(prefix[i - 1] + A[i])

        if N == B:
            return prefix[N - 1]

        # case - 1
        maxValue = prefix[N - 1] - prefix[N - 1 - B]

        # case - 2 / 3 / 4
        for L in range(0, B):
            sumValue = prefix[L] + (prefix[N - 1] - prefix[N - B + L])
            if sumValue > maxValue:
                maxValue = sumValue

        return maxValue
        
A = [5, -2, 3 , 1, 2]
B = 3

# A = [1, 2]
# B = 1

A = [ -969, -948, 350, 150, -59, 724, 966, 430, 107, -809, -993, 337, 457, -713, 753, -617, -55, -91, -791, 758, -779, -412, -578, -54, 506, 30, -587, 168, -100, -409, -238, 655, 410, -641, 624, -463, 548, -517, 595, -959, 602, -650, -709, -164, 374, 20, -404, -979, 348, 199, 668, -516, -719, -266, -947, 999, -582, 938, -100, 788, -873, -533, 728, -107, -352, -517, 807, -579, -690, -383, -187, 514, -691, 616, -65, 451, -400, 249, -481, 556, -202, -697, -776, 8, 844, -391, -11, -298, 195, -515, 93, -657, -477, 587 ]
B = 81
soln = Solution()
print(soln.solve(A, B))

File: Java-code/TopicRevision/prefixSum/pickFromBothSides.java
/*
 * First initially pick all the three balls from left side.
 */

import java.util.ArrayList;
import java.util.Arrays;

public class pickFromBothSides {
    public int solveOLD(ArrayList<Integer> A, int B) {
        int N = A.size();
        int sumVal = 0;
        ArrayList<Integer> prefix = new ArrayList<Integer>();
        prefix.add(0, A.get(0));
        for (int i = 1; i < N; i++) {
            sumVal = prefix.get(i - 1) + A.get(i);
            prefix.add(i, sumVal);
        }

        if (B == N) {
            return prefix.get(N - 1);
        }
        int maxVal = 0;
        for (int k = 0; k < B; k++) {
            maxVal = prefix.get(B - 1);
        }

        System.out.println("total left => " + maxVal);

        int p = B - 1;
        int q = 1;
        int total = 0;
        while (p >= 0 && q <= B) {
            int left = 0;
            if (p > 0) {
                left = prefix.get(p - 1);
            }

            int right = 0;
            if (q <= B) {
                right = prefix.get(N - 1) - prefix.get(N - q - 1);
            }
            total = left + right;
            if (maxVal < total) {
                maxVal = total;
            }
            System.out.println(p + " " + q);
            System.out
                    .println("left => " + left + " right => " + right + " total => " + total + " maxVal => " + maxVal);
            p--;
            q++;
        }

        return maxVal;
    }

    public int solve(ArrayList<Integer> A, int B) {
        int N = A.size();
        int prefix[] = new int[N];
        prefix[0] = A.get(0);
        for (int i = 1; i < N; i++) {
            prefix[i] = prefix[i - 1] + A.get(i);
        }

        System.out.println(Arrays.toString(prefix));

        if (B == N) {
            return prefix[N - 1];
        }

        int maxVal = Integer.MIN_VALUE;
        for (int i = 0; i <= B; i++) {
            int left = 0;
            if (i != 0) {
                left = prefix[i - 1];
            }

            int right = 0;
            int j = B - i;
            if (j != N - 1) {
                right += prefix[N - 1] - prefix[N - j - 1];
            }
            System.out.println("index i = " + (i-1) + " " + left + " || index j = " + (N - j - 1) + " " + right);

            int val = left + right;
            maxVal = Integer.max(val, maxVal);
        }
        System.out.println(maxVal);

        return maxVal;
    }

    public static void main(String[] args) {
        pickFromBothSides pk = new pickFromBothSides();

        // ArrayList<Integer> A = new ArrayList<Integer>(Arrays.asList(5, -2, 3, 1, 2));
        // int B = 3;

        ArrayList<Integer> A = new ArrayList<Integer>(Arrays.asList(2,3,-1,4,2,1));
        int B = 4;
        pk.solve(A, B);
    }
}
